### ISSUE TITLE
Crate in alternative registry breaking transitive dependency resolution for normal crates.io dependencies

### ISSUE THREAD

**Issue by campbellC** [rust-lang/cargo#9880]
**Problem**
I have a crate A in a private registry R. It depends on crates.io dependencies like semver. When I depend on crate A in a third crate B, I get an error saying that semver is not found in registry R. Cargo should be resolving this dependency in crates.io, not the private registry.

**Steps**
1. Publish a crate to a private registry, depending on standard crates.io crates
2. Depend on this private crate
3. cargo build

**Possible Solution(s)**

**Notes**

Output of `cargo version`:
cargo 1.54.0 (5ae8d74b3 2021-06-22)

Contents of Cargo.toml for crate A:
```
[package]
name = "crate-a"
version = "0.1.1"
edition = "2018"
readme = "README.md"
description = "Modify version of rust crates in place"
license = "privatecompany"
keywords = ["cli", "build-tools"]
categories = ["cli"]
repository = "https://bitbucket.privatecompany.local/projects/CORE/repos/crate-a/browse"
publish = ["cargo-corelib-local"]

[dependencies]
toml_edit = "^0.2.1"
argh = "^0.1.5"
semver = "^1.0.4"
anyhow = "^1.0.43"

[dev-dependencies]
assert_cmd = "^2.0.0"
tempfile = "^3.2.0"

[[bin]]
name = "crate-a"
```

This crate builds and I've put it up in cargo-corelib-local
Contents of Cargo.toml for crate-b
```
[package]
name = "crate-b"
version = "0.1.2"
edition = "2018"
publish = ["cargo-corelib-local"]
readme = "readme.md"
keywords = ["example"]
repository = "https://bitbucket.privatecompany.local/projects/BIG/repos/crate-b/browse"
description = "Example package"
categories = ["example"]
license = "privatecompany"

[dependencies]
rand = "0.8.4"
crate-a = { version = "0.1.0", registry = "cargo-corelib-local" } 
```
And contents of ~/.cargo/config.toml
```
[registries]
cargo-corelib-local = { index = "https://artifactory.privatecompany.local/artifactory/git/cargo-corelib-local.git" }

[net]
git-fetch-with-cli = true
```

Here's the build output from building crate-b

```
    Updating `https://artifactory.privatecompany.local/artifactory/git/cargo-corelib-local.git` index
error: no matching package named `semver` found
location searched: registry `https://artifactory.privatecompany.local/artifactory/git/cargo-corelib-local.git`
required by package `crate-a v0.1.0 (registry `https://artifactory.privatecompany.local/artifactory/git/cargo-corelib-local.git`)`
    ... which is depended on by `crate-b v0.1.2 (/home/username/Developer/crate-b)`
```

So you can see it is correctly pulling in crate-a and knows to find semver, but is incorrectly looking in the cargo-corelib-local registry rather than crates.io.

**Comment by ehuss** [comment]
This should be working.  Can you check the index files?  The `"deps"` array should have `"registry": "https://github.com/rust-lang/crates.io-index"`.  If it doesn't, it sounds like a bug in your registry implementation?

**Comment by campbellC** [comment]
Sorry I’ve not debugged registry stuff before. How can I view the index files?

**Comment by ehuss** [comment]
The git repository at https://artifactory.privatecompany.local/artifactory/git/cargo-corelib-local.git contains the index.  The files are split on the beginning of the package name (like `ca/rg/cargo` is the file for "cargo").  Inside is JSON, one object per line, where each object is a published version for that package.  Take a look at the JSON and in the "deps" array, which lists the dependencies for that package, it should have several fields, and one of them is "registry" which indicates what registry the dependency is from.  If it is null (or missing), then it means the dependency comes from the same index as where the files are located.

**Comment by campbellC** [comment]
Ok yes I can see that and indeed the registry field is missing. Ok this should be a bug report for the registry implementor in that case - apologies for the misfile. Thanks for the support!

**Comment by josh-zozonz** [comment]
@campbellC did you get this resolved with arfifactory?

**Comment by campbellC** [comment]
If I’m not mistaken, this issue was caused by a much larger problem with Artifactory’s Cargo registry implementation. Artifactory’s stored all its crate metadata in a table and the size of the data was too big for the column. So it just silently dropped lots of data from dependencies. 

This was supposedly fixed in release 7.35.1 under issue RTFACT-26495 but I haven’t tested it since my solution was to workaround using artifactory entirely and just use git.

**Comment by atoptsoglou** [comment]
what solved the issue in our end was that the user that was publishing crates did not have the annotate permission.
